Different advection schemes in different directions

simone-silvestri opened the pull request:
This was already possible for tracers with the `TracerAdvection` scheme. This PR changes the `TracerAdvection` scheme into a `FluxFormAdvection` scheme that can be used also for momentum

closes #3699
closes #3622

glwagner commented:
Does this close #3622?

glwagner commented on src/Advection/flux_form_advection.jl:
```diff
@@ -0,0 +1,52 @@
+    z :: C
+
+    FluxFormAdvection{N, FT}(x::A, y::B, z::C) where {N, FT, A, B, C} = new{N, FT, A, B, C}(x, y, z)
+end
```
Isn't this identical to `TracerAdvection`? I don't think we need `TracerAdvection` anymore.

glwagner commented on src/Advection/momentum_advection_operators.jl:
```diff
@@ -89,3 +89,20 @@ which ends up at the location `ccf`.
+##### Fallback advection fluxes!
+#####
+
+for flux_dir in (:Uu, :Vu, :Wu, :Uv, :Vv, :Wv, :Uw, :Vw, :Ww)
+    advective_momentum_flux = Symbol(:_advective_momentum_flux_, flux_dir)
```
```suggestion
for flux in (:Uu, :Vu, :Wu, :Uv, :Vv, :Wv, :Uw, :Vw, :Ww)
    advective_momentum_flux = Symbol(:_advective_momentum_flux_, flux)
```

glwagner commented:
Using this branch I get

```julia
julia> using Oceananigans
Precompiling Oceananigans
  1 dependency successfully precompiled in 11 seconds. 139 already precompiled.

julia> grid = RectilinearGrid(size = (16, 1, 16),
                              halo = (3, 1, 3),
                              x = (0, 1),
                              y = (0, 1),
                              z = (0, 1),
                              topology = (Periodic, Periodic, Bounded))
16×1×16 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 3×1×3 halo
├── Periodic x ∈ [0.0, 1.0) regularly spaced with Δx=0.0625
├── Periodic y ∈ [0.0, 1.0) regularly spaced with Δy=1.0
└── Bounded  z ∈ [0.0, 1.0] regularly spaced with Δz=0.0625

julia> model = NonhydrostaticModel(; grid, advection = WENO(order=5))
┌ Warning: Inflating model grid halo size to (3, 3, 3) and recreating grid. Note that an ImmersedBoundaryGrid requires an extra halo point in all non-flat directions compared to a non-immersed boundary grid.
└ @ Oceananigans.Models.NonhydrostaticModels ~/Projects/Oceananigans.jl/src/Models/NonhydrostaticModels/nonhydrostatic_model.jl:245
ERROR: ArgumentError: halo must be ≤ size for coordinate y
Stacktrace:
 [1] validate_halo(TX::Type, TY::Type, TZ::Type, size::Tuple{Int64, Int64, Int64}, halo::Tuple{Int64, Int64, Int64})
   @ Oceananigans.Grids ~/Projects/Oceananigans.jl/src/Grids/input_validation.jl:87
 [2] validate_rectilinear_grid_args(topology::Tuple{…}, size::Tuple{…}, halo::Tuple{…}, FT::Type, extent::Nothing, x::Tuple{…}, y::Tuple{…}, z::Tuple{…})
   @ Oceananigans.Grids ~/Projects/Oceananigans.jl/src/Grids/rectilinear_grid.jl:292
 [3] RectilinearGrid(architecture::CPU, FT::DataType; size::Tuple{…}, x::Tuple{…}, y::Tuple{…}, z::Tuple{…}, halo::Tuple{…}, extent::Nothing, topology::Tuple{…})
   @ Oceananigans.Grids ~/Projects/Oceananigans.jl/src/Grids/rectilinear_grid.jl:269
 [4] with_halo(new_halo::Tuple{…}, old_grid::RectilinearGrid{…})
   @ Oceananigans.Grids ~/Projects/Oceananigans.jl/src/Grids/rectilinear_grid.jl:389
 [5] inflate_grid_halo_size(::RectilinearGrid{…}, ::WENO{…}, ::Vararg{…})
   @ Oceananigans.Models.NonhydrostaticModels ~/Projects/Oceananigans.jl/src/Models/NonhydrostaticModels/nonhydrostatic_model.jl:250
 [6] NonhydrostaticModel(; grid::RectilinearGrid{…}, clock::Clock{…}, advection::WENO{…}, buoyancy::Nothing, coriolis::Nothing, stokes_drift::Nothing, forcing::@NamedTuple{}, closure::Nothing, boundary_conditions::@NamedTuple{}, tracers::Tuple{}, timestepper::Symbol, background_fields::@NamedTuple{}, particles::Nothing, biogeochemistry::Nothing, velocities::Nothing, hydrostatic_pressure_anomaly::Oceananigans.Models.NonhydrostaticModels.DefaultHydrostaticPressureAnomaly, nonhydrostatic_pressure::Field{…}, diffusivity_fields::Nothing, pressure_solver::Nothing, immersed_boundary::Nothing, auxiliary_fields::@NamedTuple{})
   @ Oceananigans.Models.NonhydrostaticModels ~/Projects/Oceananigans.jl/src/Models/NonhydrostaticModels/nonhydrostatic_model.jl:181
 [7] top-level scope
   @ REPL[3]:1
Some type information was truncated. Use `show(err)` to see complete types.
```

simone-silvestri commented:
you cannot use `WENO(; order =5)` in the y-direction if you have only one grid point. 
In this case the script would have to change to 
```julia
julia> using Oceananigans
Precompiling Oceananigans
  1 dependency successfully precompiled in 11 seconds. 139 already precompiled.

julia> grid = RectilinearGrid(size = (16, 1, 16),
                              halo = (3, 1, 3),
                              x = (0, 1),
                              y = (0, 1),
                              z = (0, 1),
                              topology = (Periodic, Periodic, Bounded))
16×1×16 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 3×1×3 halo
├── Periodic x ∈ [0.0, 1.0) regularly spaced with Δx=0.0625
├── Periodic y ∈ [0.0, 1.0) regularly spaced with Δy=1.0
└── Bounded  z ∈ [0.0, 1.0] regularly spaced with Δz=0.0625

julia> advection = FluxFormAdvection(WENO(order=5), nothing, WENO(order=5))

julia> model = NonhydrostaticModel(; grid, advection)
```
I think spitting out the error should be the correct behavior because we want to make sure that people know what scheme is begin used in the different directions, and correct the advection scheme accordingly. We can probably change the error message to be a bit more descriptive

glwagner commented:
i think we should support that syntax, its unambiguous in my opinion what is supposed to happen

glwagner commented:
This also fails:

```julia
julia> grid = RectilinearGrid(size = (16, 2, 16),
                              halo = (3, 2, 3),
                              x = (0, 1),
                              y = (0, 1),
                              z = (0, 1),
                              topology = (Periodic, Periodic, Bounded))
16×2×16 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 3×2×3 halo
├── Periodic x ∈ [0.0, 1.0) regularly spaced with Δx=0.0625
├── Periodic y ∈ [0.0, 1.0) regularly spaced with Δy=0.5
└── Bounded  z ∈ [0.0, 1.0] regularly spaced with Δz=0.0625

julia> model = NonhydrostaticModel(; grid, advection = WENO())
┌ Warning: Inflating model grid halo size to (3, 3, 3) and recreating grid. Note that an ImmersedBoundaryGrid requires an extra halo point in all non-flat directions compared to a non-immersed boundary grid.
└ @ Oceananigans.Models.NonhydrostaticModels ~/Projects/Oceananigans.jl/src/Models/NonhydrostaticModels/nonhydrostatic_model.jl:245
ERROR: ArgumentError: halo must be ≤ size for coordinate y
Stacktrace:
 [1] validate_halo(TX::Type, TY::Type, TZ::Type, size::Tuple{Int64, Int64, Int64}, halo::Tuple{Int64, Int64, Int64})
   @ Oceananigans.Grids ~/Projects/Oceananigans.jl/src/Grids/input_validation.jl:87
 [2] validate_rectilinear_grid_args(topology::Tuple{…}, size::Tuple{…}, halo::Tuple{…}, FT::Type, extent::Nothing, x::Tuple{…}, y::Tuple{…}, z::Tuple{…})
   @ Oceananigans.Grids ~/Projects/Oceananigans.jl/src/Grids/rectilinear_grid.jl:292
 [3] RectilinearGrid(architecture::CPU, FT::DataType; size::Tuple{…}, x::Tuple{…}, y::Tuple{…}, z::Tuple{…}, halo::Tuple{…}, extent::Nothing, topology::Tuple{…})
   @ Oceananigans.Grids ~/Projects/Oceananigans.jl/src/Grids/rectilinear_grid.jl:269
 [4] with_halo(new_halo::Tuple{…}, old_grid::RectilinearGrid{…})
   @ Oceananigans.Grids ~/Projects/Oceananigans.jl/src/Grids/rectilinear_grid.jl:389
 [5] inflate_grid_halo_size(::RectilinearGrid{…}, ::WENO{…}, ::Vararg{…})
   @ Oceananigans.Models.NonhydrostaticModels ~/Projects/Oceananigans.jl/src/Models/NonhydrostaticModels/nonhydrostatic_model.jl:250
 [6] NonhydrostaticModel(; grid::RectilinearGrid{…}, clock::Clock{…}, advection::WENO{…}, buoyancy::Nothing, coriolis::Nothing, stokes_drift::Nothing, forcing::@NamedTuple{}, closure::Nothing, boundary_conditions::@NamedTuple{}, tracers::Tuple{}, timestepper::Symbol, background_fields::@NamedTuple{}, particles::Nothing, biogeochemistry::Nothing, velocities::Nothing, hydrostatic_pressure_anomaly::Oceananigans.Models.NonhydrostaticModels.DefaultHydrostaticPressureAnomaly, nonhydrostatic_pressure::Field{…}, diffusivity_fields::Nothing, pressure_solver::Nothing, immersed_boundary::Nothing, auxiliary_fields::@NamedTuple{})
   @ Oceananigans.Models.NonhydrostaticModels ~/Projects/Oceananigans.jl/src/Models/NonhydrostaticModels/nonhydrostatic_model.jl:181
 [7] top-level scope
   @ REPL[6]:1
Some type information was truncated. Use `show(err)` to see complete types.
```

ie when we don't specify an order at all for `WENO`. Why should this fail? It looks perfectly fine, reading the code.

simone-silvestri commented:
@glwagner after the last commit the code should have the expected behavior of reducing the advection order to comply with grid-size limitations. I am not doing the same check for the vector invariant scheme because it might get a bit complicated.

simone-silvestri commented:
This now spits out
```julia
julia> grid = RectilinearGrid(size = (16, 2, 16),
                                     halo = (3, 1, 3),
                                     x = (0, 1),
                                     y = (0, 1),
                                     z = (0, 1),
                                     topology = (Periodic, Periodic, Bounded))
16×2×16 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 3×1×3 halo
├── Periodic x ∈ [0.0, 1.0) regularly spaced with Δx=0.0625
├── Periodic y ∈ [0.0, 1.0) regularly spaced with Δy=0.5
└── Bounded  z ∈ [0.0, 1.0] regularly spaced with Δz=0.0625

julia> model = NonhydrostaticModel(; grid, advection = WENO());
[ Info: User-defined advection scheme WENO reconstruction order 5 reduced to WENO reconstruction order 3 in the y-direction to comply with grid-size limitations.
┌ Warning: Inflating model grid halo size to (3, 2, 3) and recreating grid. Note that an ImmersedBoundaryGrid requires an extra halo point in all non-flat directions compared to a non-immersed boundary grid.
└ @ Oceananigans.Models.NonhydrostaticModels ~/development/Oceananigans.jl/src/Models/NonhydrostaticModels/nonhydrostatic_model.jl:250

julia> grid = RectilinearGrid(size = (16, 1, 16),
                                     halo = (3, 1, 3),
                                     x = (0, 1),
                                     y = (0, 1),
                                     z = (0, 1),
                                     topology = (Periodic, Periodic, Bounded))
16×1×16 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 3×1×3 halo
├── Periodic x ∈ [0.0, 1.0) regularly spaced with Δx=0.0625
├── Periodic y ∈ [0.0, 1.0) regularly spaced with Δy=1.0
└── Bounded  z ∈ [0.0, 1.0] regularly spaced with Δz=0.0625

julia> model = NonhydrostaticModel(; grid, advection = WENO());
[ Info: User-defined advection scheme WENO reconstruction order 5 reduced to Nothing in the y-direction to comply with grid-size limitations.
```

glwagner commented on src/Advection/adapt_advection_order.jl:
```diff
@@ -0,0 +1,120 @@
+function adapt_advection_order(advection::FluxFormAdvection, grid::AbstractGrid)
+    advection_x = adapt_advection_order_x(advection.x, grid, grid.Nx)
+    advection_y = adapt_advection_order_y(advection.y, grid, grid.Ny)
+    advection_z = adapt_advection_order_z(advection.z, grid, grid.Nz)
```
does

```julia
adapt_advection_order(advection.z, topology(grid, 3), size(grid, 3), grid)
```

work?

simone-silvestri commented:
It does now. Although the topology argument does nothing for the moment.

glwagner commented on src/Advection/adapt_advection_order.jl:
```diff
@@ -39,9 +41,9 @@ function adapt_advection_order(advection, grid::AbstractGrid)
-    advection_z = adapt_advection_order_z(advection.z, grid, grid.Nz)
+    advection_x = adapt_advection_order_x(advection.x, topology(grid, 1), size(grid, 1), grid)
+    advection_y = adapt_advection_order_y(advection.y, topology(grid, 2), size(grid, 2), grid)
+    advection_z = adapt_advection_order_z(advection.z, topology(grid, 3), size(grid, 3), grid)
```
And now you don't need _x, _y, _z right?

simone-silvestri replied:
We need it because of WENO schemes that might have stretched directions in x, y or z so we have to check if we need to reconstruct the scheme using grid information or not.

simone-silvestri replied:
probably we can just check that any direction is stretched. Then we use the grid.

glwagner replied:
I see, it's not a topology thing

simone-silvestri commented:
I have incurred into a problem with the enzyme. 

There was an issue with the `N` parameter of `ScalarDiffusivity` and `ScalarBiharmonicDiffusivity` (which tells us how large the halo size has to be). 
I fixed this problem in this PR by switching a couple of parameters around (I need to add a unit test for that before this PR is merged), but now the constructor is not type-stable anymore so the enzyme tests fail. 
I have changed the constructor to accept `Val(halo_size_required)` instead of an integer and this fixes the tests, but I am not completely convinced of this solution (Why would we need to pass `Val(Int)` instead of an integer?).

What would the options be here? Completely change where we store the buffer size (quite annoying because it is nice to abstract it away in the parameters of the type)? Is there a way to make the constructor type stable without having to do this hacky change to the UI?

simone-silvestri commented:
I have found a solution. Just forcing the keyword argument `required_halo_size` with to be an integer `required_halo_size::Int =1` stabilized the constructor. I need to add a couple of tests and this PR should be ready

simone-silvestri commented:
Also, do we need a type-stable constructor for `ScalarDiffusivity`? Basically all constructors we have are not type stable so I don't see why ScalarDiffusivity should be type-stable. If we want the diffusivity to be mutable (like in the enzyme test) we can pass an array or a field to the `ScalarDiffusivity` instead of reconstructing the diffusivity type.

simone-silvestri commented:
So I have reverted back to the previous UI (added a couple of unit tests) and fixed the tests for enzyme to use Field diffusivities instead of integers that are not mutable

glwagner commented on src/Models/NonhydrostaticModels/compute_nonhydrostatic_boundary_tendencies.jl:
```diff
@@ -64,15 +64,16 @@ end
-    Oxᴿ = (Nx-B,  0, 0)
-    Oyᴿ = (0,  Ny-B, 0)
+    Oxᴿ = (Nx-Bx,  0, 0)
+    Oyᴿ = (0,  Ny-By, 0)
```
might be clearer to switch to the new `KernelParameters` syntax

glwagner commented on src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_model.jl:
```diff
@@ -130,6 +130,12 @@ function HydrostaticFreeSurfaceModel(; grid,
 
     tracers = tupleit(tracers) # supports tracers=:c keyword argument (for example)
 
+      # Adjust advection scheme to be valid on a particular grid size. i.e. if the grid size
```
```suggestion
    # Adjust advection scheme to be valid on a particular grid size. i.e. if the grid size
```

glwagner commented on src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_model.jl:
```diff
@@ -130,6 +130,12 @@ function HydrostaticFreeSurfaceModel(; grid,
 
     tracers = tupleit(tracers) # supports tracers=:c keyword argument (for example)
 
+      # Adjust advection scheme to be valid on a particular grid size. i.e. if the grid size
+    # is smaller than the advection order, reduce the order of the advection in that particular
+    # direction
```
```suggestion
    # Reduce the advection order in directions where reconstruction is limited by the number or grid points
```

glwagner commented on src/Advection/adapt_advection_order.jl:
```diff
@@ -0,0 +1,111 @@
+using Oceananigans.Grids: topology
+
+"""
+    adapt_advection_order(advection, grid::AbstractGrid)
```
This could just be `validate_advection_scheme(scheme, grid)`. We often refer to this kind of thing as a "validation"

simone-silvestri replied:
I tend to think at a validation as a non-returning function

glwagner commented:
Can you briefly explain why this PR is absolutely _huge_ when it seems to do something quite modest?

glwagner commented on src/Advection/adapt_advection_order.jl:
```diff
@@ -0,0 +1,111 @@
+end
+
+function adapt_advection_order(advection::UpwindBiased{H}, N::Int, grid::AbstractGrid) where H
+    if N >= H
```
isn't `H` called `B` elsewhere? Confused because this does not seem like a halo

glwagner commented on src/Advection/adapt_advection_order.jl:
```diff
@@ -0,0 +1,111 @@
+    changed_advection = any((changed_x, changed_y, changed_z))
+
+    if changed_x
+        @info "User-defined advection scheme $(summary(advection)) reduced to $(summary(new_advection.x)) in the x-direction to comply with grid-size limitations."
```
```suggestion
        @info "Using $(summary(new_advection.x)) in the x-direction because size(grid, 1) = $(size(grid, 1))"
```

glwagner replied:
same for the rest

glwagner commented on src/Advection/momentum_advection_operators.jl:
```diff
@@ -89,3 +89,97 @@ which ends up at the location `ccf`.
+@inline _advective_momentum_flux_Ww(i, j, k, grid, ::Nothing, W, ::ZeroField) = zero(grid)
+@inline _advective_momentum_flux_Wu(i, j, k, grid, ::Nothing, ::ZeroField, u) = zero(grid)
+@inline _advective_momentum_flux_Wv(i, j, k, grid, ::Nothing, ::ZeroField, v) = zero(grid)
+@inline _advective_momentum_flux_Ww(i, j, k, grid, ::Nothing, ::ZeroField, w) = zero(grid)
```
dang

glwagner commented on src/Advection/momentum_advection_operators.jl:
```diff
@@ -89,3 +89,97 @@ which ends up at the location `ccf`.
+
+        @inline _advective_momentum_flux_Wu(i, j, k, grid, ::$Scheme, ::ZeroField, u) = zero(grid)
+        @inline _advective_momentum_flux_Wv(i, j, k, grid, ::$Scheme, ::ZeroField, v) = zero(grid)
+        @inline _advective_momentum_flux_Ww(i, j, k, grid, ::$Scheme, ::ZeroField, w) = zero(grid)
```
This loop creates 27 * 4 = 108 functions that return 0. Why

glwagner commented on src/Advection/tracer_advection_operators.jl:
```diff
@@ -61,20 +24,24 @@ Adapt.adapt_structure(to, scheme::TracerAdvection{N, FT}) where {N, FT} =
+        # Fallback for `ZeroField` velocities
+        @inline _advective_tracer_flux_x(i, j, k, grid, ::$Scheme, ::ZeroField, c) = zero(grid)
+        @inline _advective_tracer_flux_y(i, j, k, grid, ::$Scheme, ::ZeroField, c) = zero(grid)
+        @inline _advective_tracer_flux_z(i, j, k, grid, ::$Scheme, ::ZeroField, c) = zero(grid)
```
Another 9 * 4 = 36

simone-silvestri commented:
> Can you briefly explain why this PR is absolutely _huge_ when it seems to do something quite modest? I guess it adds 450 lines. But it's actually effectively more because of the metaprogramming. Just wondering if we are not thinking critically about this

Right, it it because the behaviour for `nothing` advective schemes and `ZeroField` tracers have been moved into the `_advective_flux` function. That is no required, let me do a couple of changes

glwagner commented:
> > Can you briefly explain why this PR is absolutely _huge_ when it seems to do something quite modest? I guess it adds 450 lines. But it's actually effectively more because of the metaprogramming. Just wondering if we are not thinking critically about this
> 
> Right, it it because the behaviour for `nothing` advective schemes and `ZeroField` tracers have been moved into the `_advective_flux` function. That is no required, let me do a couple of changes

I trust your judgement provided you have thought critically about how to ensure the implementation is as concise as possible (and noting in comments in the code where improvements are to be made). We want the source code to be as short as possible.

simone-silvestri commented:
This should be ready to go

glwagner commented on src/Advection/adapt_advection_order.jl:
```diff
@@ -0,0 +1,111 @@
+end
+
+function adapt_advection_order(advection::FluxFormAdvection, grid::AbstractGrid)
+    advection_x = adapt_advection_order(advection.x, size(grid, 1), grid)
+    advection_y = adapt_advection_order(advection.y, size(grid, 2), grid)
+    advection_z = adapt_advection_order(advection.z, size(grid, 3), grid)
```
```suggestion
    original_x_advection = x_advection(advection)
    original_y_advection = y_advection(advection)
    original_z_advection = z_advection(advection)

    adapted_x_advection = adapt_advection_order(original_advection, size(grid, 1), grid)
    adapted_y_advection = adapt_advection_order(original_advection, size(grid, 2), grid)
    adapted_z_advection = adapt_advection_order(original_advection, size(grid, 3), grid)
```

You just need

```julia
x_advection(flux_form::FluxFormAdvection) = flux_form.x
x_advection(advection) = advection
```

and then you can avoid all the copy-pasting

glwagner commented on src/Advection/adapt_advection_order.jl:
```diff
@@ -0,0 +1,111 @@
+    if N >= B
+        return advection
+    else
+        return new_weno_scheme(advection, grid, N * 2 - 1, advection.bounds, XT, YT, ZT)
```
```suggestion
        return new_weno_scheme(advection, grid, 2N - 1, advection.bounds, XT, YT, ZT)
```

glwagner commented on src/Advection/adapt_advection_order.jl:
```diff
@@ -0,0 +1,111 @@
+    if N >= B
+        return advection
+    else
+        return UpwindBiased(; order = N * 2 - 1)
```
```suggestion
        return UpwindBiased(; order = 2N - 1)
```

glwagner commented on src/Advection/adapt_advection_order.jl:
```diff
@@ -0,0 +1,111 @@
+    if N >= B
+        return advection
+    else
+        return Centered(; order = N * 2)
```
```suggestion
        return Centered(; order = 2N)
```

glwagner commented on test/test_enzyme.jl:
```diff
@@ -16,10 +16,8 @@ const maximum_diffusivity = 100
-    names = tuple(:c) # tracernames(model.tracers)
-    closure = with_tracers(names, closure)
-    model.closure = closure
+    closure = model.closure
+    fill!(closure.κ.c, diffusivity)
```
Can you change it back? We wanted this specific test (a parameter estimation-type test)

simone-silvestri replied:
see comment below

glwagner commented on test/test_enzyme.jl:
```diff
@@ -169,7 +167,10 @@ end
 
     grid = RectilinearGrid(size=(Nx, Ny, Nz); x, y, z, topology)
-    diffusion = VerticalScalarDiffusivity(κ=0.1)
+    κ = ZFaceField(grid)
+    fill!(κ, 0.1)
+
+    diffusion = VerticalScalarDiffusivity(; κ)
```
Please restore this code to how it was!

glwagner replied:
I think it would be better to mark this as `@test_broken` then. We need the constructor to be type stable to do this work!

glwagner replied:
Please understand what I am saying: the point of this test is not to get a numerical result. It's supposed to verify that we can use this method for doing parameter estimation with this specific setup. So if the code breaks then the test fails.

simone-silvestri replied:
ok I can do that

glwagner replied:
> There was a problem with this test. The constructor of the ScalarDiffusivity is not type-stable (like most constructors)

Many constructors are type stable, and we have to work to make them all type stable to support AD.

This test passed previously, so not sure what changed in this PR?

glwagner commented on src/TurbulenceClosures/turbulence_closure_implementations/scalar_diffusivity.jl:
```diff
@@ -219,3 +220,25 @@ function on_architecture(to, closure::ScalarDiffusivity{TD, F, <:Any, <:Any, N})
 end
+
+# We do not have at the moment the ability to distinguish between the halos required in the x, y and z
+# direction for a general closure since the closure could contain any user-defined function
```
I don't really understand this comment. We can easily tell if a closure has a function in it or not?

glwagner commented on src/TurbulenceClosures/turbulence_closure_implementations/scalar_biharmonic_diffusivity.jl:
```diff
@@ -117,3 +120,25 @@ function on_architecture(to, closure::ScalarBiharmonicDiffusivity{F, <:Any, <:An
+#= 
+required_halo_size_x(::ScalarBiharmonicDiffusivity{<:Any, ThreeDimensionalFormulation, N}) where N = N
+required_halo_size_y(::ScalarBiharmonicDiffusivity{<:Any, ThreeDimensionalFormulation, N}) where N = N
+required_halo_size_z(::ScalarBiharmonicDiffusivity{<:Any, ThreeDimensionalFormulation, N}) where N = N
```
This code may go stale cause there's no way to test it so we may not want to include a comment like this

glwagner reviewed (approved):
Please restore the enzyme tests!

glwagner commented:
> This should be ready to go

Please just restore the enzyme tests because we would like to maintain a parameter-estimation style test. For a closure like CATKE, we have to use the pattern where we set `model.closure = new_closure`. So despite that using an array for the diffusivity is _functionally_ equivalent to what was written, it is definitely not equivalent in terms of the code that is exercised for taking a gradient.
